TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs

```diff
@@ -88,59 +123,161 @@ private void UpdateRoad()
+                        int junctionIndex = _path.GetClosestIndexOnPath(anchor1);
+                        
+                        // Set the start and end indices accordingly
+                        startIndex = isStart ? edgeIndex : junctionIndex;
+                        endIndex = isStart ? junctionIndex : edgeIndex;
+                    }
+                    
+                }
+                else if(intersection.Type == IntersectionType.FourWayIntersection)
+                {
+                    Vector3 anchor1 = intersection.Road1 == this ? intersection.Road1AnchorPoint1 : intersection.Road2AnchorPoint1;
+                    Vector3 anchor2 = intersection.Road1 == this ? intersection.Road1AnchorPoint2 : intersection.Road2AnchorPoint2;
+                    startIndex = _path.GetClosestIndexOnPath(anchor1);
+                    endIndex = _path.GetClosestIndexOnPath(anchor2);
+                }
+
+                Vector3 intersectionPoint = intersection.IntersectionPosition;
+                intersectionVertices.Enqueue(new IntersectionVertexPoints(startIndex, intersectionPoint, endIndex, intersection.Type, intersection.Road1 == this));
             }
 
             // Go through each point in the path of the road
-            for(int i = 1; i < _path.NumPoints; i++)
+            for(int i = 0; i < _path.NumPoints; i++)
             {
                 // Add an intersection node if there is an intersection between the previous node and the current node
-                for (int j = 0; j < distanceAtIntersection.Count; j++)
+                IntersectionVertexPoints? possibleNextIntersection = intersectionVertices.Count > 0 ? intersectionVertices.Peek() : null;
+                if(possibleNextIntersection != null)
                 {
-                    // If the path distance at the intersection is between the previous node distance and the current node distance
-                    if (distanceAtIntersection[j] > _path.cumulativeLengthAtEachVertex[i-1] && distanceAtIntersection[j] < _path.cumulativeLengthAtEachVertex[i])
+                    IntersectionVertexPoints nextIntersection = (IntersectionVertexPoints)possibleNextIntersection;
+                    if(i == nextIntersection.StartIndex || i == nextIntersection.EndIndex)
                     {
-                        Vector3 intersectionPosition = _intersections[j].IntersectionPosition;
+                        // If the intersection is a 3-way intersection, we want to add a junction node:
+                        // To both indices if this is Road1
+                        // To the end index if if the intersection is at the start
+                        // To the start index if the intersection is at the end
+                        bool threeWayShouldAddJunctionEdge = 
+                            nextIntersection.IsThreeWayRoad1
+                            || (nextIntersection.Type == IntersectionType.ThreeWayIntersectionAtStart && i == nextIntersection.EndIndex)
```

Comment:
could make this and the line below into its own bools to increase readability

Reply:
I don't really think that will improve the readability. It is just three simple checks and with the comment explaining the logic I don't think it will be easier to read by splitting it

This logic is already its own bool to separate it from the four way intersection logic and improve the readability in the if statement